tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp

```diff
@@ -533,3 +533,91 @@ namespace result_test {
     STATIC_ASSERT(test_convertible_3<in_in_out_result>());
     STATIC_ASSERT(test_convertible_3<in_out_out_result>());
 } // namespace result_test
+
+namespace permutable_test {
+    using std::forward_iterator, std::indirectly_movable_storable, std::indirectly_swappable, std::permutable;
+
+    template <int I>
+    struct archetype {
+        using value_type      = int;
+        using difference_type = int;
+
+        struct proxy {
+            proxy()             = default;
+            proxy(proxy const&) = delete;
+            proxy& operator=(proxy const&) = delete;
+
+            operator int() const;
+            // 1: not reflexively indirectly_movable_storable
+            void operator=(int) const requires(I != 1);
+        };
+
+        proxy operator*() const;
+
+        archetype& operator++();
+        archetype operator++(int);
+
+        // 0: not forward_iterator (input only)
+        bool operator==(archetype const&) const requires(I != 0) = default;
+
+        friend int iter_move(archetype const&) {
+            return 42;
+        }
+        friend void iter_swap(archetype const&, archetype const&) {}
+
+        // Ideally we'd have a "not reflexively indirectly_swappable" case as well, but there's no way for a class to
+        // satisfy indirectly_movable_storable<T, T> without satisfying indirectly_swappable<T, T> thanks to N4861
+        // [iterator.cust.swap]/4.3. permutable requires indirectly_swappable<T, T> only to forbid a user type to define
+        // an iter_swap overload that doesn't meet the semantic requirements.
+    };
+    STATIC_ASSERT(std::input_iterator<archetype<0>>);
+    STATIC_ASSERT(!std::forward_iterator<archetype<0>>);
+    STATIC_ASSERT(indirectly_movable_storable<archetype<0>, archetype<0>>);
+    STATIC_ASSERT(indirectly_swappable<archetype<0>, archetype<0>>);

```

Comment:
Line 538 says:
```cpp
using std::forward_iterator, std::indirectly_movable_storable, std::indirectly_swappable, std::permutable;
```

`std::forward_iterator` is qualified here, which seems inconsistent. Occurs below.